08-optional-types.Rmd

```diff
@@ -0,0 +1,344 @@
+
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+
+    # these work as you would expect
+    Boolean did_x_run = defined(task_x.something)
+    Boolean did_y_run = defined(task_y.something)
+
+    outputs {
+        File? task_x_output = task_x.something
+        File? task_y_output = task_y.something
+        Boolean did_x_run = did_x_run
+        Boolean did_y_run = did_y_run
+    }
+}
+```
+
+Be aware that WDL is not "aware" of any of these checks. Even if your workflow has established a variable is defined, WDL does not know that, so you cannot use these methods to use an optional variable like a non-optional variable.
+
+```
+if defined(some_integer) {
+    Int real_some_integer = some_integer
+}
+```
+Instead, you need to either use your variables as optional variables, or coerce them into their respective not-optional type.
+
+You should not use `defined()` with optional arrays, as it often acts in unexpected ways. It is very easy to accidentally create a `defined()` check that always returns true, as WDL executors may define "optional" arrays even if they have no values. For example, let's consider task_x and task_y again. To recap, these are mutually exclusive tasks which each output a single File. If these tasks are scattered from outside their respective if-blocks, the their outputs are no longer File?, instead, they are arrays. But are they arrays of optional File?s, or are the arrays themselves optional? Let's try to find out:
+
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        Array[File] some_inputs
+    }
+    scatter(some_input in some_inputs) {
+        if (do_x = true) {
+            call task_x {
+                input:
+                    some_input = some_input
+            }
+        }
+        if (do_x = false) {
+            call task_y {
+            input:
+                    some_input = some_input
+            }
+        }
+    }
+
+    # these both always return true!
+    Boolean did_x_run = defined(task_x.something)
+    Boolean did_y_run = defined(task_y.something)
+
+    outputs {
+        File? task_x_output = task_x.something
+        File? task_y_output = task_y.something
+        Boolean did_x_run = did_x_run
+        Boolean did_y_run = did_y_run
+    }
+}
+```
+
+Instead of using `defined()` on an array, consider simply checking its length. An empty array has a length of 0, so if the length of an array is greater than 1, you know that there is at least one item existing in that array.
+
+## Coercing arrays with optional components
```

Comment:
These last two sections are quite technical and I'm not sure if everyone can appreciate this immediately. I wonder if we should put this into a tucked-away section.